libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -820,56 +1128,60 @@ struct __single_thread_group {
     constexpr _CUDA_VSTD::size_t thread_rank() const { return 0; };
 };
 
-template<typename _Group, class _Tp, thread_scope _Sco, typename _CompF>
+template<typename _Group, class _Tp, typename _Size, thread_scope _Sco, typename _CompF>
 _LIBCUDACXX_INLINE_VISIBILITY
-async_contract_fulfillment memcpy_async(_Group const & __group, _Tp * __destination, _Tp const * __source, _CUDA_VSTD::size_t __size, barrier<_Sco, _CompF> & __barrier) {
-    // When compiling with NVCC and GCC 4.8, certain user defined types that _are_ trivially copyable are
-    // incorrectly classified as not trivially copyable. Remove this assertion to allow for their usage with
-    // memcpy_async when compiling with GCC 4.8.
-    // FIXME: remove the #if once GCC 4.8 is no longer supported.
-#if !defined(_LIBCUDACXX_COMPILER_GCC) || _GNUC_VER > 408
-    static_assert(_CUDA_VSTD::is_trivially_copyable<_Tp>::value, "memcpy_async requires a trivially copyable type");
-#endif
-
-    return __memcpy_async<alignof(_Tp)>(__group, reinterpret_cast<char *>(__destination), reinterpret_cast<char const *>(__source), __size, __barrier);
+async_contract_fulfillment __memcpy_async_barrier(_Group const & __group, _Tp * __destination, _Tp const * __source, _Size __size, barrier<_Sco, _CompF> & __barrier) {
+    // 1. Determine which completion mechanisms can be used with the current
+    // barrier. A local shared memory barrier, i.e., block-scope barrier in local
+    // shared memory, supports the mbarrier_complete_tx mechanism in addition to
+    // the async group mechanism.
+    _CUDA_VSTD::uint32_t __allowed_completions = __is_local_smem_barrier(__barrier)
+        ? ( _CUDA_VSTD::uint32_t(__completion_mechanism::__async_group) | _CUDA_VSTD::uint32_t(__completion_mechanism::__mbarrier_complete_tx))
+        : _CUDA_VSTD::uint32_t(__completion_mechanism::__async_group);
+
+    // 2. Issue actual copy instructions.
+    auto __bh = __try_get_barrier_handle(__barrier);
+    auto __cm =  __dispatch_memcpy_async(__single_thread_group{}, __destination, __source, __size, __allowed_completions, __bh);
+
+    // 3. Synchronize barrier with copy instructions.
+    return __memcpy_completion_impl::__delay(__cm, __group, __size, __barrier);
```

Comment:
I really don't like the word "delay" for this. At least for me, when I see "delay", I think a timed delay, not a delay until another thing has happened.